File: server/src/lib/store/mcp.ts
import { join } from "path";
import { homedir } from "os";
import { existsSync, mkdirSync } from "fs";
import { readJsonc, writeJsonc } from "../jsonc.js";
import * as oc from "../config/opencode.js";
import * as mc from "../config/mimocode.js";
import * as cl from "../config/claude.js";
import type { McpServer, McpType } from "@carrot-switch/shared";

const STORE_ROOT = join(homedir(), "AppData", "Roaming", ".carrotswitch");

function storePath(agent: string): string {
  return join(STORE_ROOT, "mcps", `${agent}.json`);
}

function nowIso(): string {
  return new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
}

function ensureDir(filePath: string): void {
  const dir = filePath.substring(0, filePath.lastIndexOf("\\") !== -1 ? filePath.lastIndexOf("\\") : filePath.lastIndexOf("/"));
  if (!existsSync(dir)) {
    mkdirSync(dir, { recursive: true });
  }
}

function getConfig(agent: string) {
  const map: Record<string, any> = { opencode: oc, mimocode: mc, claude: cl };
  return map[agent];
}

export interface McpStoreData {
  servers: Record<string, McpServer>;
}

export function load(agent: string): McpStoreData {
  const path = storePath(agent);
  if (!existsSync(path)) {
    return seedFromAgent(agent);
  }
  const data = readJsonc(path) as McpStoreData;
  for (const server of Object.values(data.servers || {})) {
    if (server.type === undefined) server.type = "local";
  }
  return data;
}

function seedFromAgent(agent: string): McpStoreData {
  const config = getConfig(agent);
  if (!config || !config.is_available()) return { servers: {} };

  const agentServers = config.get_mcp_servers();
  const data: McpStoreData = { servers: {} };
  for (const [name, server] of Object.entries(agentServers)) {
    const s = server as Record<string, any>;
    data.servers[name] = {
      name,
      type: (s.type || "local") as McpType,
      command: s.command,
      url: s.url,
      environment: s.environment,
      enabled: s.enabled !== false,
      createdAt: nowIso(),
    };
  }
  save(agent, data);
  return data;
}

export function save(agent: string, data: McpStoreData): void {
  const path = storePath(agent);
  ensureDir(path);
  writeJsonc(path, data);
}

export function addServer(agent: string, name: string, server: Omit<McpServer, "createdAt"> & { createdAt?: string }): void {
  const data = load(agent);
  data.servers[name] = {
    ...server,
    name,
    createdAt: nowIso(),
    enabled: server.enabled !== false,
  };
  save(agent, data);
}

export function updateServer(agent: string, name: string, server: Partial<McpServer>): void {
  const data = load(agent);
  if (!data.servers || !(name in data.servers)) {
    throw new Error(`MCP server '${name}' not found`);
  }
  const existing = data.servers[name];
  data.servers[name] = {
    ...existing,
    ...server,
    name,
    createdAt: existing.createdAt || nowIso(),
    enabled: server.enabled !== undefined ? server.enabled : existing.enabled,
  };
  save(agent, data);
}

export function deleteServer(agent: string, name: string): void {
  const data = load(agent);
  if (!data.servers || !(name in data.servers)) {
    throw new Error(`MCP server '${name}' not found`);
  }
  delete data.servers[name];
  save(agent, data);
}

export function toggleServer(agent: string, name: string): boolean {
  const data = load(agent);
  if (!data.servers || !(name in data.servers)) {
    throw new Error(`MCP server '${name}' not found`);
  }
  const current = data.servers[name].enabled !== false;
  data.servers[name].enabled = !current;
  save(agent, data);
  return !current;
}

export function syncToAgent(agent: string): void {
  const config = getConfig(agent);
  if (!config) return;

  const data = load(agent);
  for (const [name, server] of Object.entries(data.servers || {})) {
    const agentServer: Record<string, any> = {};
    if (server.type) agentServer.type = server.type;
    if (server.command) agentServer.command = server.command;
    if (server.url) agentServer.url = server.url;
    if (server.environment) agentServer.environment = server.environment;
    try {
      config.update_mcp_server(name, agentServer);
    } catch {
      config.add_mcp_server(name, agentServer);
    }
  }
}

File: src/carrot_switch/store/mcp.py
"""MCP server local store — primary source of truth."""
from pathlib import Path
from carrot_switch.store import STORE_ROOT, _ensure_dir, _now_iso
from carrot_switch.config import read_jsonc, write_jsonc


def _store_path(agent: str) -> Path:
    return STORE_ROOT / "mcps" / f"{agent}.json"


def load(agent: str) -> dict:
    """Load local MCP store. Returns {"servers": {...}}."""
    path = _store_path(agent)
    if not path.exists():
        return _seed_from_agent(agent)
    data = read_jsonc(path)
    # Ensure every server has a type field
    for server in data.get("servers", {}).values():
        server.setdefault("type", "local")
    return data


def _seed_from_agent(agent: str) -> dict:
    """First run: import from agent config file."""
    from carrot_switch.config import opencode as oc, mimocode as mc, claude as cl

    if agent == "opencode":
        config = oc
    elif agent == "mimocode":
        config = mc
    elif agent == "claude":
        config = cl
    else:
        return {"servers": {}}

    if not config.is_available():
        return {"servers": {}}

    agent_servers = config.get_mcp_servers()
    data = {"servers": {}}
    for name, server in agent_servers.items():
        data["servers"][name] = {
            **server,
            "enabled": server.get("enabled", True),
            "created_at": _now_iso(),
        }
    save(agent, data)
    return data


def save(agent: str, data: dict) -> None:
    """Persist MCP store to disk."""
    _ensure_dir(_store_path(agent))
    write_jsonc(_store_path(agent), data)


def add_server(agent: str, name: str, server: dict) -> None:
    data = load(agent)
    server["created_at"] = _now_iso()
    if "enabled" not in server:
        server["enabled"] = True
    data["servers"][name] = server
    save(agent, data)


def update_server(agent: str, name: str, server: dict) -> None:
    data = load(agent)
    if name not in data["servers"]:
        raise KeyError(f"MCP server '{name}' not found")
    server["created_at"] = data["servers"][name].get("created_at", _now_iso())
    if "enabled" not in server:
        server["enabled"] = data["servers"][name].get("enabled", True)
    data["servers"][name] = server
    save(agent, data)


def delete_server(agent: str, name: str) -> None:
    data = load(agent)
    if name not in data["servers"]:
        raise KeyError(f"MCP server '{name}' not found")
    del data["servers"][name]
    save(agent, data)


def toggle_server(agent: str, name: str) -> bool:
    data = load(agent)
    if name not in data["servers"]:
        raise KeyError(f"MCP server '{name}' not found")
    current = data["servers"][name].get("enabled", True)
    data["servers"][name]["enabled"] = not current
    save(agent, data)
    return not current


def sync_to_agent(agent: str) -> None:
    """Write-through: push local store state to agent config file.

    Strips local-only fields (enabled, created_at) from what goes to agent config.
    """
    from carrot_switch.config import opencode as oc, mimocode as mc, claude as cl

    if agent == "opencode":
        config = oc
    elif agent == "mimocode":
        config = mc
    elif agent == "claude":
        config = cl
    else:
        return

    data = load(agent)
    for name, server in data["servers"].items():
        agent_server = {k: v for k, v in server.items() if k not in ("enabled", "created_at")}
        try:
            config.update_mcp_server(name, agent_server)
        except KeyError:
            config.add_mcp_server(name, agent_server)
